### ISSUE TITLE
NaN Issues During Multi-GPU Training

### ISSUE THREAD

**Issue by liuxin99** [yzhangcs/ctc-copy#2]
Hello,

I'm very interested in your work and have been trying to replicate it. However, I'm encountering an issue where the training consistently results in NaN values. I am using the default learning rate settings as specified in your documentation. My training setup includes 4 x 24G NVIDIA TITAN RTX GPUs on a slurm cluster.

Could you please help me identify any potential reasons for this issue? Below is my `train.sh` script for reference:
```
# nohup bash train.sh > log 2>&1 &

module purge
module add anaconda3 cuda-11.7
conda activate ctc-copy
export CUDA_LAUNCH_BLOCKING=1
args=$@
for arg in $args; do
    eval "$arg"
done

echo "seed:        ${seed:=1}"
echo "bert:        ${bert:=roberta-large}"
echo "lr1:         ${lr1:=5e-5}"
echo "lr2:         ${lr2:=5e-6}"
echo "lr3:         ${lr3:=1e-6}"
echo "rate1:       ${rate1:=10}"
echo "rate2:       ${rate2:=10}"
echo "rate3:       ${rate3:=10}"
echo "upsampling:  ${upsampling:=4}"
echo "batch:       ${batch:=80000}"
echo "epochs1:     ${epochs1:=64}"
echo "epochs2:     ${epochs2:=64}"
echo "epochs3:     ${epochs3:=64}"
echo "warmup1:     ${warmup1:=1000}"
echo "warmup2:     ${warmup2:=0}"
echo "warmup3:     ${warmup3:=0}"
echo "glat:        ${glat:=1}"
echo "update:      ${update:=10}"
echo "devices:     ${devices:=0,1,2,3}"
echo "config:      ${config:=configs/roberta.yaml}"
echo "Path:        ${Path:=exp/ctc.roberta}"

code=$Path.code
mkdir -p $Path.code
cp run.py $code/
cp -r ctc $code/
cp -r 3rdparty $code/
printf "Current commits:\n$(git log -1 --oneline)\n3rd parties:\n"
cd 3rdparty/parser/ && printf "parser\n$(git log -1 --oneline)\n" && cd ../..

for stage in 1; do
    mkdir -p $Path/stage$stage
    # var="lr$stage";     lr=${!var}
    # var="rate$stage";   rate=${!var}
    # var="warmup$stage"; warmup=${!var}
    # var="epochs$stage"; epochs=${!var}
    if [ "$stage" -eq 1 ]; then
        lr=$lr1
        rate=$rate1
        warmup=$warmup1
        epochs=$epochs1
    elif [ "$stage" -eq 2 ]; then
        lr=$lr2
        rate=$rate2
        warmup=$warmup2
        epochs=$epochs2
    elif [ "$stage" -eq 3 ]; then
        lr=$lr3
        rate=$rate3
        warmup=$warmup3
        epochs=$epochs3
    fi
    current="$Path/stage$stage/model.lr$lr.rate$rate.upsampling$upsampling.batch$batch.epochs$epochs.warmup$warmup.glat$glat.seed$seed"

    if [ $stage -eq 1 ]; then
        train=data/clang8.train
        (set -x
        python -u run.py train -b -s $seed -d $devices -c $config -p $current --lr=$lr  --lr-rate=$rate  --upsampling=$upsampling --batch-size=$batch --epochs=$epochs  --warmup-steps=$warmup  --glat=$glat --update-steps=$update --encoder=bert --bert=$bert --train $train --eval-tgt --cache --amp --wandb
        )
    else
        if [ $stage -eq 2 ]; then
            train=data/error_coded.train
        else
            train=data/wi_locness.train
        fi
        (set -x
        cp $prev $current
        python -u run.py train -s $seed -d $devices -c $config -p $current --lr=$lr  --lr-rate=$rate  --upsampling=$upsampling --batch-size=$batch --epochs=$epochs  --warmup-steps=$warmup  --glat=$glat --update-steps=$update --encoder=bert --bert=$bert --train $train --eval-tgt --cache
        )
    fi
    # zsh pred.sh Path=$current
    # prev=$current
done
```
And the following pic is the wandb log of lr and loss:
<img width="740" alt="image" src="https://github.com/yzhangcs/ctc-copy/assets/33060143/b2603bbc-5b95-4eee-be1c-d3ea1559e2d6">

**Comment by yzhangcs** [comment]
@liuxin99 Hi, not sure what exactly happened, is it convenient for you to paste your full training logs or send by email?

**Comment by liuxin99** [comment]
Ok, I'll send my full training logs by email. Thank you very much.
